backend/app/models.py

```diff
@@ -1,6 +1,325 @@
+
+        if not self.part_of_speech_tags:
+            text = nltk.word_tokenize(self.text)
+            pos_tags = nltk.pos_tag(text)
+
+            self.part_of_speech_tags = pos_tags
+            self.save()
+        return self.part_of_speech_tags
+
+    def get_part_of_speech_words(self, words, remove_swords=True):
+        """
+        A helper method for retrieving the number of occurrences of input words keyed to their
+        NLTK tag values (i.e., 'NN' for noun).
+
+        :param words: a list of strings.
+        :param remove_swords: optional boolean, remove stop words from return.
+        :return: a dictionary keying NLTK tag strings to Counter instances.
+        """
+        stop_words = set(nltk.corpus.stopwords.words('english'))
+        document_pos_tags = self.get_part_of_speech_tags()
+        words_set = {word.lower() for word in words}
+        output = {}
+
+        for token, tag in document_pos_tags:
+            lowered_token = token.lower()
+            if remove_swords is True and token in stop_words:
+                continue
+            if token not in words_set:
+                continue
+            if tag not in output:
+                output[tag] = Counter()
+            output[tag][lowered_token] += 1
+
+        return output
+
+    def update_metadata(self, new_metadata):
+        """
+        Updates the metadata of the document without requiring a complete reloading
+        of the text and other properties.
+
+        'filename' cannot be updated with this method.
+
+        :param new_metadata: dict of new metadata to apply to the document
+        :return: None
+        """
+        default_fields = [field.name for field in self._meta.get_fields()]
+        if not isinstance(new_metadata, dict):
+            raise ValueError(
+                f'new_metadata must be a dictionary of metadata keys, not type {type(new_metadata)}'
+            )
+
+        for key in new_metadata:
+            if key == 'date':
+                try:
+                    new_metadata[key] = int(new_metadata[key])
+                except ValueError as err:
+                    raise ValueError(
+                        f"the metadata field 'date' must be a number for the document,"
+                        f" not '{new_metadata['date']}'"
+                    ) from err
+            if key not in default_fields:
+                self.new_attributes[key] = new_metadata[key]
+            else:
+                setattr(self, key, new_metadata[key])
+
+        if 'text' in new_metadata or not self.tokenized_text:
+            self.text = self._clean_quotes()
+            self.word_count = None
+            self.tokenized_text = None
+            self.word_counts_counter = dict()
+            self.part_of_speech_tags = list()
+            self.save()
+            # get_word_count() will call get_tokenized_text() and update both
+            # self.word_count and self.tokenized_text
+            self.get_word_count()
```

Comment:
The need for a comment here seems like a smell to me. In the ideal world, we'd be able to be more explicit about what's happening in this method so that we aren't requiring the presence of either an explanatory comment or the subsequent developer to go hunting through other method implementations. i.e., something like:
```
self.get_tokenized_text()
self.get_word_count()
etc.
```

In general, this points to one of the challenges of porting over the existing implementation. All of these methods were constructed to memoize when required (more or less) rather than perform some of the expensive operations up front. That pattern made sense when they instances were being stored in local memory, but I think we could move to a pattern where we perform all of the expensive operations up front when a `Document` row is created.